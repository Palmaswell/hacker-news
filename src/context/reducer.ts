import { Story } from '.';

export interface ReducerState {
  readonly ids: number[];
  readonly published: Story[];
  readonly stories: Story[];
  readonly counter: number;
}

export interface Action {
  readonly type: ActionType;
  readonly ids?: number[];
  readonly story?: Story;
  readonly stories?: Story[];
  readonly counter?: number;
}

export interface InitFetchStories {
  readonly ids: number[];
  readonly counter: number;
  publish(story: Story): void;
  push(story: Story): void;
}

export enum ActionType {
  fetchIds = 'fetchIds',
  push = 'push',
  pushBulk = 'pushBulk',
  publish = 'publish',
  publishBulk = 'publishBulk',
  setCounter = 'setCounter',
}

export function publisher(
  stories: Story[],
  published: Story[],
  counter: number,
): Story[] | [] {
  const lastPublished = published[published.length - 1];
  const isPublished = stories
    .slice(published.length)
    .find(story => story.id === lastPublished.id);

  if (isPublished) {
    return [];
  }

  return published.concat(
    stories.slice(published.length, published.length + counter),
  );
}

export async function fetchIds(): Promise<number[]> {
  try {
    const res = await fetch(
      'https://hacker-news.firebaseio.com/v0/newstories.json',
    );
    const ids = await res.json();

    return ids;
  } catch (error) {
    console.warn(`Fetching story ids failed with: ${error}`);

    return Promise.resolve([]);
  }
}

export async function fetchStories({
  ids,
  counter,
  publish,
  push,
}: InitFetchStories): Promise<Story[]> {
  try {
    return await Promise.all(
      ids.map(async (id: number, idx: number) => {
        const res = await fetch(
          `https://hacker-news.firebaseio.com/v0/item/${id}.json`,
        );
        const hackerStory = await res.json();

        const story = {
          by: hackerStory.by,
          id: hackerStory.id,
          time: hackerStory.time,
          title: hackerStory.title,
          url: hackerStory.url,
        };
        if (idx <= counter) {
          publish(story);
        }
        if (idx <= counter * 4) {
          push(story);
        }

        return story;
      }),
    );
  } catch (error) {
    console.warn(`Fetch stories failed with: ${error}`);
    return Promise.resolve([]);
  }
}

export async function* collectStories(
  ids: number[],
): AsyncGenerator<Story, void, Story> {
  for await (const id of ids) {
    const res = await fetch(
      `https://hacker-news.firebaseio.com/v0/item/${id}.json`,
    );
    const hackerStory = await res.json();
    const story = {
      by: hackerStory.by,
      id: hackerStory.id,
      time: hackerStory.time,
      title: hackerStory.title,
      url: hackerStory.url,
    };

    yield story;
  }
}

export function reducer(state: ReducerState, action: Action): ReducerState {
  switch (action.type) {
    case ActionType.fetchIds: {
      return {
        ...state,
        ...(Array.isArray(action.ids) && { ids: [...action.ids] }),
      };
    }

    case ActionType.publish: {
      return {
        ...state,
        ...(action.story && { published: [...state.published, action.story] }),
      };
    }

    case ActionType.publishBulk: {
      return {
        ...state,
        ...(action.stories && {
          published: [...action.stories],
        }),
      };
    }

    case ActionType.push: {
      return {
        ...state,
        ...(action.story && { stories: [...state.stories, action.story] }),
      };
    }

    case ActionType.pushBulk: {
      return {
        ...state,
        ...(action.stories && {
          stories: [...action.stories],
        }),
      };
    }

    case ActionType.setCounter: {
      return {
        ...state,
        ...(action.counter && { counter: action.counter }),
      };
    }
    default: {
      throw new Error(`Unhandled type: ${action.type}`);
    }
  }
}
