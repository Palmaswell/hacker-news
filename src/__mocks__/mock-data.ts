import { Story } from '../context';

export const storiesMock: Story[] = [
  {
    by: 'makeworld',
    id: 23957535,
    time: '1595775730',
    title: 'Crypto 101',
    url: 'https://www.crypto101.io/',
  },
  {
    by: 'farseek',
    id: 23957534,
    time: '1595775724',
    title: 'Big ‘Ol List of Jobs',
    url: 'https://www.crypto101.io/',
  },
  {
    by: 'Lemmih',
    id: 23957530,
    time: '1595775705',
    title: 'Voice Controlled Animations',
    url: 'https://reanimate.readthedocs.io/en/latest/voice/',
  },
  {
    by: 'based2',
    id: 23957527,
    time: '1595775677',
    title: 'Apache Camel 2.15.4 Released',
    url:
      'http://mail-archives.apache.org/mod_mbox/www-annou…6qHYBrJPR006RtkPLRyVGs2uRQJ3A%40mail.gmail.com%3E',
  },
  {
    by: 'based2',
    id: 23957510,
    time: '1595775536',
    title: 'New ‘Meow’ attack has deleted almost 4k unsecured databases',
    url:
      'https://www.bleepingcomputer.com/news/security/new…ack-has-deleted-almost-4-000-unsecured-databases/',
  },
  {
    by: 'pseudolus',
    id: 23957505,
    time: '1595775507',
    title: 'What’s Going on Inside Argentina’s Fearsome Thunderstorms?',
    url:
      'https://www.nytimes.com/interactive/2020/07/22/magazine/worst-storms-argentina.html',
  },
  {
    by: 'shahinrostami',
    id: 23957485,
    time: '1595775301',
    title:
      'Show HN: I made an interactive explorer for co-occurring Pokémon Types',
    url:
      'https://shahinrostami.com/posts/statistics/data-is-beautiful/co-occurring-pokemon-types/',
  },
  {
    by: 'verandaguy',
    id: 23957483,
    time: '1595775293',
    title: 'Overview of the Contoso Corporation',
    url:
      'https://docs.microsoft.com/en-us/microsoft-365/enterprise/contoso-overview?view=o365-worldwide',
  },
  {
    by: 'evo_9',
    id: 23957459,
    time: '1595775113',
    title:
      '20 years ago, Steve Jobs built the “coolest computer ever”–and it bombed',
    url:
      'https://arstechnica.com/gadgets/2020/07/20-years-a…bs-built-the-coolest-computer-ever-and-it-bombed/',
  },
  {
    by: 'teddyh',
    id: 23957455,
    time: '1595775086',
    title: 'Headphones are collecting too much personal data',
    url:
      'https://www.soundguys.com/headphones-are-collecting-too-much-personal-data-21524/',
  },
  {
    by: 'Fiveplus',
    id: 23957437,
    time: '1595774937',
    title:
      'A former chaos engineer offers tips for handling online disasters remotely',
    url:
      'https://techcrunch.com/2020/04/01/a-former-chaos-e…rs-5-tips-for-handling-online-disasters-remotely/',
  },
  {
    by: 'afrcnc',
    id: 23957430,
    time: '1595774899',
    title: 'Dramatic decline found in Western bumblebee populations',
    url:
      'https://www.timesunion.com/news/article/Dramatic-decline-found-in-Western-bumblebee-15433741.php',
  },
  {
    by: 'sai18',
    id: 23957424,
    time: '1595774820',
    title: 'Logify: Log out of all your online accounts with one click',
    url: 'https://www.crypto101.io/',
  },
  {
    by: 'raviksharma',
    id: 23957388,
    time: '1595774575',
    title: 'Bartosz Milewski – Basics of Haskell',
    url: 'https://github.com/raviksharma/bartosz-basics-of-haskell',
  },
  {
    by: 'markshepard',
    id: 23957381,
    time: '1595774546',
    title: 'Employers are forcing employees  to take time off to avoid burnout',
    url:
      'https://www.cnbc.com/2020/07/26/avoiding-coronavir…urnout-by-forcing-employees-to-take-time-off.html',
  },
  {
    by: 'fortran77',
    id: 23957376,
    time: '1595774526',
    title: 'IEEE Top Programming Languages 2020',
    url:
      'https://spectrum.ieee.org/at-work/tech-careers/top-programming-language-2020',
  },
  {
    by: 'bequanna',
    id: 23957373,
    time: '1595774484',
    title: 'Will Garmin Pay $10M Ransom to End Two-Day Outage?',
    url:
      'https://www.forbes.com/sites/barrycollins/2020/07/…will-garmin-pay-10m-ransom-to-end-two-day-outage/',
  },
  {
    by: 'whack',
    id: 23957365,
    time: '1595774435',
    title:
      'Research on psychedelics and their effects in alcoholism, major depressio (cont)',
    url:
      'https://www.psychologytoday.com/us/blog/balanced/202007/magic-mushrooms-and-the-future-psychology',
  },
  {
    by: 'tomohawk',
    id: 23957350,
    time: '1595774278',
    title:
      'Seattle ‘Housing Affordability’ Law Forces Hard-Working Homeowners to Pay Ransom',
    url:
      'https://ij.org/press-release/seattle-housing-affor…lding-permit-now-two-residents-are-fighting-back/',
  },
];

export const publishedMock: Story[] = [
  {
    by: 'makeworld',
    id: 23957535,
    time: '1595775730',
    title: 'Crypto 101',
    url: 'https://www.crypto101.io/',
  },
  {
    by: 'farseek',
    id: 23957534,
    time: '1595775724',
    title: 'Big ‘Ol List of Jobs',
    url: 'https://www.crypto101.io/',
  },
  {
    by: 'Lemmih',
    id: 23957530,
    time: '1595775705',
    title: 'Voice Controlled Animations',
    url: 'https://reanimate.readthedocs.io/en/latest/voice/',
  },
  {
    by: 'based2',
    id: 23957527,
    time: '1595775677',
    title: 'Apache Camel 2.15.4 Released',
    url:
      'http://mail-archives.apache.org/mod_mbox/www-annou…6qHYBrJPR006RtkPLRyVGs2uRQJ3A%40mail.gmail.com%3E',
  },
  {
    by: 'based2',
    id: 23957510,
    time: '1595775536',
    title: 'New ‘Meow’ attack has deleted almost 4k unsecured databases',
    url:
      'https://www.bleepingcomputer.com/news/security/new…ack-has-deleted-almost-4-000-unsecured-databases/',
  },
  {
    by: 'pseudolus',
    id: 23957505,
    time: '1595775507',
    title: 'What’s Going on Inside Argentina’s Fearsome Thunderstorms?',
    url:
      'https://www.nytimes.com/interactive/2020/07/22/magazine/worst-storms-argentina.html',
  },
];
