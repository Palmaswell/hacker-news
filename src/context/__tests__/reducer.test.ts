import { createStory, publisher } from '../reducer';
import { storiesMock, publishedMock } from '../../__mocks__/mock-data';
import { Story } from '..';

describe('Reducer', () => {
  describe('publisher()', () => {
    it('should return an empty array when a story has already been published', () => {
      const published = [...publishedMock];
      published.pop();
      published.push(storiesMock[publishedMock.length]);

      expect(publisher(storiesMock, published, 6)).toEqual([]);
    });

    it('should increment the amount of published stories on each call according to counter', () => {
      const counter = 6;
      const initialBulk = publisher(storiesMock, publishedMock, counter);
      expect(initialBulk.length).toBe(counter * 2);
      const secondBulk = publisher(storiesMock, initialBulk, counter);
      expect(secondBulk.length).toBe(counter * 3);
    });

    it('should NOT publish more stories than cached stories', () => {
      const totalStories = storiesMock.length;
      const counter = 4;
      publishedMock.reduce((acc: Story[], _: Story, idx: number): Story[] => {
        if (idx < 2) {
          return [...acc];
        }
        const bulk = publisher(storiesMock, acc, counter);

        expect(bulk.length <= totalStories).toBeTruthy();
        return [...bulk];
      }, publishedMock);
    });

    it('should return an array containing the next bulk of stories to be published', () => {
      expect(publisher(storiesMock, publishedMock, 6)).toMatchSnapshot();
    });
  });

  describe('createStory()', () => {
    let initialTime: number;
    let date: string;
    let currentDate: string;
    beforeEach(() => {
      initialTime = 1595859421;
      date = new Date(initialTime * 1000).toLocaleString();
      currentDate = new Date(Date.now() * 1000).toLocaleString();
    });

    afterEach(() => {
      initialTime = (undefined as unknown) as number;
      date = (undefined as unknown) as string;
      currentDate = (undefined as unknown) as string;
    });
    it('should return null on missing story', () => {
      expect(createStory(null)).toBeNull();
    });

    it('should return a string containing unknown on missing by property', () => {
      expect(
        createStory({
          descendants: 0,
          id: 23965288,
          score: 1,
          time: initialTime,
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          type: 'story',
          url: 'https://name/',
        }),
      ).toEqual(
        expect.objectContaining({
          by: 'unknown',
          id: 23965288,
          time: date,
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          url: 'https://name/',
        }),
      );
    });

    it('should return a string containing unknown on missing title property', () => {
      expect(
        createStory({
          by: 'robertwiblin',
          descendants: 0,
          id: 23965288,
          score: 1,
          time: initialTime,
          type: 'story',
          url: 'https://name/',
        }),
      ).toEqual(
        expect.objectContaining({
          by: 'robertwiblin',
          id: 23965288,
          time: date,
          title: 'unknown',
          url: 'https://name/',
        }),
      );
    });

    it('should return a string containing unknown on missing url property', () => {
      expect(
        createStory({
          by: 'robertwiblin',
          descendants: 0,
          id: 23965288,
          score: 1,
          time: initialTime,
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          type: 'story',
        }),
      ).toEqual(
        expect.objectContaining({
          by: 'robertwiblin',
          id: 23965288,
          time: date,
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          url: 'unknown',
        }),
      );
    });

    it('should return the current Date on missing time property', () => {
      expect(
        createStory({
          by: 'robertwiblin',
          descendants: 0,
          id: 23965288,
          score: 1,
          url: 'https://name/',
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          type: 'story',
        }),
      ).toEqual(
        expect.objectContaining({
          by: 'robertwiblin',
          id: 23965288,
          time: currentDate,
          title: 'Ben Garfinkel on scrutinising classic AI risk arguments',
          url: 'https://name/',
        }),
      );
    });
  });
});
