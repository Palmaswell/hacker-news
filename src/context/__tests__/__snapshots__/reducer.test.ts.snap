// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer publisher() should return an array containing the next bulk of stories to be published 1`] = `
Array [
  Object {
    "by": "makeworld",
    "id": 23957535,
    "time": "1595775730",
    "title": "Crypto 101",
    "url": "https://www.crypto101.io/",
  },
  Object {
    "by": "farseek",
    "id": 23957534,
    "time": "1595775724",
    "title": "Big ‘Ol List of Jobs",
    "url": "https://www.crypto101.io/",
  },
  Object {
    "by": "Lemmih",
    "id": 23957530,
    "time": "1595775705",
    "title": "Voice Controlled Animations",
    "url": "https://reanimate.readthedocs.io/en/latest/voice/",
  },
  Object {
    "by": "based2",
    "id": 23957527,
    "time": "1595775677",
    "title": "Apache Camel 2.15.4 Released",
    "url": "http://mail-archives.apache.org/mod_mbox/www-annou…6qHYBrJPR006RtkPLRyVGs2uRQJ3A%40mail.gmail.com%3E",
  },
  Object {
    "by": "based2",
    "id": 23957510,
    "time": "1595775536",
    "title": "New ‘Meow’ attack has deleted almost 4k unsecured databases",
    "url": "https://www.bleepingcomputer.com/news/security/new…ack-has-deleted-almost-4-000-unsecured-databases/",
  },
  Object {
    "by": "pseudolus",
    "id": 23957505,
    "time": "1595775507",
    "title": "What’s Going on Inside Argentina’s Fearsome Thunderstorms?",
    "url": "https://www.nytimes.com/interactive/2020/07/22/magazine/worst-storms-argentina.html",
  },
  Object {
    "by": "shahinrostami",
    "id": 23957485,
    "time": "1595775301",
    "title": "Show HN: I made an interactive explorer for co-occurring Pokémon Types",
    "url": "https://shahinrostami.com/posts/statistics/data-is-beautiful/co-occurring-pokemon-types/",
  },
  Object {
    "by": "verandaguy",
    "id": 23957483,
    "time": "1595775293",
    "title": "Overview of the Contoso Corporation",
    "url": "https://docs.microsoft.com/en-us/microsoft-365/enterprise/contoso-overview?view=o365-worldwide",
  },
  Object {
    "by": "evo_9",
    "id": 23957459,
    "time": "1595775113",
    "title": "20 years ago, Steve Jobs built the “coolest computer ever”–and it bombed",
    "url": "https://arstechnica.com/gadgets/2020/07/20-years-a…bs-built-the-coolest-computer-ever-and-it-bombed/",
  },
  Object {
    "by": "teddyh",
    "id": 23957455,
    "time": "1595775086",
    "title": "Headphones are collecting too much personal data",
    "url": "https://www.soundguys.com/headphones-are-collecting-too-much-personal-data-21524/",
  },
  Object {
    "by": "Fiveplus",
    "id": 23957437,
    "time": "1595774937",
    "title": "A former chaos engineer offers tips for handling online disasters remotely",
    "url": "https://techcrunch.com/2020/04/01/a-former-chaos-e…rs-5-tips-for-handling-online-disasters-remotely/",
  },
  Object {
    "by": "afrcnc",
    "id": 23957430,
    "time": "1595774899",
    "title": "Dramatic decline found in Western bumblebee populations",
    "url": "https://www.timesunion.com/news/article/Dramatic-decline-found-in-Western-bumblebee-15433741.php",
  },
]
`;
